src/rust/engine/fs/store/src/snapshot.rs

```diff
@@ -559,6 +559,86 @@ impl Snapshot {
       })
       .to_boxed()
   }
+
+  pub fn get_snapshot_subset(
+    store: Store,
+    digest: Digest,
+    path_globs: PathGlobs,
+    workunit_store: WorkUnitStore,
+  ) -> BoxFuture<Snapshot, String> {
+    use bazel_protos::remote_execution::{Directory, DirectoryNode, FileNode};
+
+    let traverser = move |_store: &Store,
+                          path_so_far: &PathBuf,
+                          _digest: Digest,
+                          directory: &Directory|
+          -> BoxFuture<Vec<PathStat>, String> {
+      let subdir_paths: Vec<PathBuf> = directory
+        .get_directories()
+        .iter()
+        .map(move |node: &DirectoryNode| path_so_far.join(node.get_name()))
+        .collect();
+
+      let file_paths: Vec<PathBuf> = directory
+        .get_files()
+        .iter()
+        .map(|node: &FileNode| path_so_far.join(node.get_name()))
+        .collect();
+
+      let filtered_subdirs = future::result(path_globs.filter(subdir_paths));
+      let filtered_files = future::result(path_globs.filter(file_paths));
+
+      filtered_subdirs
+        .join(filtered_files)
+        .and_then(|(subdirs, files): (Vec<PathBuf>, Vec<PathBuf>)| {
+          let mut path_stats: Vec<PathStat> = vec![];
+
+          for path in subdirs.into_iter() {
+            path_stats.push(PathStat::dir(path.clone(), Dir(path)));
+          }
+
+          for path in files.into_iter() {
+            path_stats.push(PathStat::file(
+              path.clone(),
+              File {
+                path,
+                is_executable: false,
+              },
+            ));
+          }
+
+          future::ok(path_stats)
+        })
+        .to_boxed()
+    };
+
+    #[derive(Clone)]
+    struct Digester {
+      digest: hashing::Digest,
+    }
+
+    impl StoreFileByDigest<String> for Digester {
+      fn store_by_digest(
+        &self,
+        _: fs::File,
+        _: WorkUnitStore,
+      ) -> BoxFuture<hashing::Digest, String> {
+        future::ok(self.digest).to_boxed()
+      }
+    }
+
+    store
+      .walk(digest, traverser, workunit_store.clone())
+      .and_then(move |path_stats_per_directory| {
+        let mut path_stats =
+          Iterator::flatten(path_stats_per_directory.into_iter().map(Vec::into_iter))
+            .collect::<Vec<_>>();
+        path_stats.sort_by(|l, r| l.path().cmp(&r.path()));
+        let digester = Digester { digest };
+        Snapshot::from_path_stats(store, &digester, path_stats, workunit_store)
```

Comment:
This is incorrect. `StoreFileByDigest` returns the `Digest` for a particular file, so it cannot have a constant implementation like this. To implement `StoreFileByDigest` you'd need a `HashMap<PathBuf, Digest>` to use the first argument to `store_by_digest` to figure out which `Digest` to return.

Reply:
@stuhood I believe I've come up with implementation that fixes this